include/llvm/IR/DTUpdater.h

```diff
@@ -0,0 +1,72 @@
+//===- DTUpdater.h - DomTree/Post DomTree Updater ---------------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DTUpdater class, which provides a uniform way to
+// update dominator related data structures.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_DTUPDATER_H
+#define LLVM_DTUPDATER_H
+
+#include "llvm/Analysis/PostDominators.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/GenericDomTree.h"
+
+namespace llvm{
+    // Maybe needs to use template to handle those class inherit from PostDomTreeBase?
+    class DTUpdater {
+    public:
+        // updateStrategy enum class
+        enum updateStrategy : unsigned char {
+            eagerly = 0,
+            deferred = 1
+        };
+        // Constructor
+        // Because most passes now use a single strategy all over one pass
+        // So we construct DTUpdater with the strategy
+        DTUpdater(updateStrategy ups_):update_DT(false),update_PDT(false),update_lazily(isLazy(ups_)){}
+        DTUpdater(DominatorTree& DT_,updateStrategy ups_):
+                DT(&DT_),update_DT(true),update_PDT(false),update_lazily(isLazy(ups_)){}
+        DTUpdater(PostDominatorTree& PDT_,updateStrategy ups_):PDT(&PDT_),update_DT(false),update_PDT(true),update_lazily(isLazy(ups_)){}
+        DTUpdater(DominatorTree& DT_,PostDominatorTree& PDT_,updateStrategy ups_):DT(&DT_),PDT(&PDT_),update_DT(true),update_PDT(true),update_lazily(isLazy(ups_)){}
+        bool hasDomTree(){return update_DT;}
+        bool hasPostDomTree(){return update_PDT;}
+        void applyUpdates(ArrayRef<DominatorTree::UpdateType> Updates);
+        void insertEdge(BasicBlock *From, BasicBlock *To);
+        void deleteEdge(BasicBlock *From, BasicBlock *To);
+        DominatorTree& getDomTree();
+        PostDominatorTree& getPostDomTree();
+        void deleteBB(BasicBlock *DelBB);
+        bool pendingDeletedBB(BasicBlock *DelBB);
+        bool pendingDT();
+        bool pendingPDT();
+        void recalculateDT(Function &F);
+        void recalculatePDT(Function &F);
+    private:
+        bool flushDelBB();
+        DominatorTree* DT= nullptr;
+        PostDominatorTree* PDT= nullptr;
+        const bool update_DT,update_PDT,update_lazily;
+        SmallVector<DominatorTree::UpdateType, 16> PendBothTreeUpdates, PendPDTUpdates;
```

Comment:
It's more readable to keep declaration of variables in separate lines.

Reply:
I think one vector is enough. Suppose we are doing lazy updates and we have never clear the update vector -- we only always append to it. In this case when you are about to apply updates to one of the trees it should be enough to keep track of where was the last update that was applied. When the updates are applied, we just have to advance the position to be at the end of the update vector.
This is of course unrealistic as nobody has to have an infinite vector, but since we don't need updates already applied to both the trees, we can forget the ones that come before the minimum position of the cursors with the last update position for DT and PDT.
As a next step, it's easy to notice that this is going to be always better or the same as keeping two vectors in terms of memory consumption, since it takes constant size to keep two indexes that keep index of the last update applied.

Reply:
I have thought about this before I use two vectors here. There isn't any issue on memory consumption in my implementation because when the pending updates of DT are flushed, if there are further updates applied and PDT isn't flushed, all updates that haven't applied to PDT will be moved into PendPDTUpdates vector. But the cons here is that 1. the PDT needs to be updated after DT is updated (or we may have three vectors here, and the project states that using DT to prune PDT updates so though a full separation of updates of two trees is preferred, we can do this in the first stage) 2. move elements by push and pop is costly.

If here I use index recording methods to pend updates. 1. a container which can pop_front() efficiently should be used when we need to forget the range before min(DT_pend_pos, PDT_pend_pos), but I found LLVM tends to use containers with little memory allocations so I am looking into which one to use (Traditionally I would like to use `deque`, but its overhead is large; 
 lists have a lot of allocations, and maybe I should move needed elements to the front and then erase to the end of the vector). 2. the DominatorTreeBase's applyUpdates function needs to have an interface to take (iterator Update.A, iterator Update.B) to eliminate copies of pendUpdates every time I need to applyUpdates.

I prefer the one using index recording, but I think I need to look more into how to address problem 1) in this method mentioned above before I code. (I will do a quick hack on this stuff) :)